zend_abstract_interface/sandbox/sandbox.h

```diff
@@ -0,0 +1,374 @@
+typedef struct zai_error_state_s {
+    int type;
+    int lineno;
+    char *message;
+    char *file;
+    int error_reporting;
+    zend_error_handling error_handling;
+} zai_error_state;
+
+typedef struct zai_exception_state_s {
+    zval *exception;
+    zval *prev_exception;
+    zend_op *opline_before_exception;
+} zai_exception_state;
+
+typedef struct zai_sandbox_s {
+    zai_error_state error_state;
+    zai_exception_state exception_state;
+} zai_sandbox;
+
+inline void zai_sandbox_error_state_backup_ex(zai_error_state *es TSRMLS_DC) {
+    es->type = PG(last_error_type);
+    es->lineno = PG(last_error_lineno);
+    es->message = PG(last_error_message);
+    es->file = PG(last_error_file);
+
+    PG(last_error_type) = 0;
+    PG(last_error_lineno) = 0;
+    /* We need to null these so that if another error comes along they do not
+     * get double-freed.
+     */
+    PG(last_error_message) = NULL;
+    PG(last_error_file) = NULL;
+
+    es->error_reporting = EG(error_reporting);
+    EG(error_reporting) = 0;
+    zend_replace_error_handling(EH_SUPPRESS, NULL, &es->error_handling TSRMLS_CC);
+}
+
+inline void zai_sandbox_error_state_restore_ex(zai_error_state *es TSRMLS_DC) {
+    if (PG(last_error_message)) {
+        if (PG(last_error_message) != es->message) {
+            free(PG(last_error_message));
+        }
+        if (PG(last_error_file) != es->file) {
+            free(PG(last_error_file));
+        }
+    }
+    zend_restore_error_handling(&es->error_handling TSRMLS_CC);
+    PG(last_error_type) = es->type;
+    PG(last_error_message) = es->message;
+    PG(last_error_file) = es->file;
+    PG(last_error_lineno) = es->lineno;
+    EG(error_reporting) = es->error_reporting;
+}
+
+inline void zai_sandbox_exception_state_backup_ex(zai_exception_state *es TSRMLS_DC) {
+    es->exception = NULL;
+    es->prev_exception = NULL;
+    if (EG(exception)) {
+        es->opline_before_exception = EG(opline_before_exception);
+        es->exception = EG(exception);
+        es->prev_exception = EG(prev_exception);
+        EG(exception) = NULL;
+        EG(prev_exception) = NULL;
+    }
+}
+
+inline void zai_sandbox_exception_state_restore_ex(zai_exception_state *es TSRMLS_DC) {
+    if (EG(exception)) {
+        /* We cannot use zend_clear_exception() here in PHP 5 since there is no
+         * NULL check on the opline.
+         */
+        zval_ptr_dtor(&EG(exception));
+        EG(exception) = NULL;
+        if (EG(prev_exception)) {
+            zval_ptr_dtor(&EG(prev_exception));
+            EG(prev_exception) = NULL;
+        }
+        if (EG(current_execute_data)) {
+            EG(current_execute_data)->opline = EG(opline_before_exception);
+        }
+    }
+
+    if (es->exception) {
+        EG(exception) = es->exception;
+        EG(prev_exception) = es->prev_exception;
+
+        EG(opline_before_exception) = es->opline_before_exception;
```

Comment:
I'm not entirely sure if I understood this properly, on PHP 7/8 you use `zend_clear_exception` which updates `EG(current_execute_data)->opline` with the one stored in `EG(opline_before_exception)` and then you call `zend_throw_exception_internal` which restores `EG(opline_before_exception)` with the one stored in `EG(current_execute_data)->opline` (so basically unchanged), wouldn't this still be the opline before the exception which occurred within the sandboxed code? If so, shouldn't you follow the same strategy and store the `opline_before_exception` in 7 and 8?

Reply:
Very nice catch! You are correct, this hidden state is not being treated the same way across all the PHP versions. And also on PHP 7+, `zend_throw_exception_hook` gets invoked an extra time when the sandbox closes because of calling `zend_throw_exception_internal`. Based on your feedback I updated this in 6cd3540 and added a test to ensure `zend_throw_exception_hook` does not get called an extra time.

I still think there is something off with how we're not backing up the original opline, but I haven't thought of a test to that would prove that this isn't working as expected yet.